posts/abcs-of-dqm-balance

```diff
@@ -0,0 +1,80 @@
+
+Events allow you to transmit data from one application to another asynchronously. Your data quality jobs can be triggered by notification or event upon completion of your ETL job or jobs. One of these would ideally be your balance job. In an event-driven ETL process, balancing your data is a little trickier than a traditional ETL process, as it is more difficult to distinguish between a job and a task. A platform that processes credit card transactions, for example, would be processing each purchase as a separate event. There are usually a series of processes a bank or credit card company uses to verify transactions and eventually post them to your account. How would you balance such a system?
+
+## An Aside on Orchestration in Event-Driven ETL
+
+In the [Balancing in Practice](#balancing-in-practice) section above, we looked at a balance architecture that ran serially; upon completion of our ETL Job, we ran the audit process, followed by the balance process, followed by control. In an event-driven ETL pipeline, the entire ETL execution could mean processing thousands or millions of events. Thus we would have to balance and control while we are completing the ETL Job execution.
+
+![Orchestration of DQM](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2020/07/abcs-dqm-balance-3.png)
+
+## Balancing the Event-Driven Way
+
+When balancing the event-driven way, we first need to define what an event looks like for the ABC Process. Events for your DQM Process will mainly be capturing the metadata created by your ETL Job. Some of the metadata that can be included in the event are:
+
+- Primary Keys
+- Data Types
+- Row Counts
+- Fields to audit for quality
+- Fields to balance for quality
+- Queries to be executed for auditing or balancing your data
+
+With this understanding of what an event looks like, we can now examine what balancing could look like.
+
+![Event-Driven Balance](https://raw.githubusercontent.com/ippontech/blog-usa/master/images/2020/07/abcs-dqm-balance-4.png)
+
+In the diagram above, I've simplified the event, which would be the input to the balance process, to three specific types of metadata: primary keys, queries to be executed, and row counts. For each event, you would want to check that all primary key constraints have been upheld to ensure data integrity. You would also want to check the row counts and execute any queries for balancing data. If all the constraints and row counts match up, much like the case discussed above, your data is balanced. However, if there are any discrepancies, the response would be different as part of the balance job. For an event-driven balance process, for each event that encounters inconsistencies as a part of the balance process, the results and even specific rows from the dataset could be sent to the control process. From here, it would again be the realm of the control process, which will be addressed in an upcoming post, to correct the data.
```

Comment:
In the diagram above, I have simplified the event, which would be the input to the balance process, into three specific types of metadata: primary keys, queries to be executed, and row counts.